Include/internal/pycore_stackref.h

```diff
@@ -0,0 +1,189 @@
+    static inline PyObject *
+    PyStackRef_Get(_PyStackRef tagged) {
+        PyObject *cleared = ((PyObject *)((tagged).bits & (~Py_TAG)));
+        return cleared;
+    }
+#else
+    #define PyStackRef_Get(tagged) ((PyObject *)(uintptr_t)((tagged).bits))
+#endif
+
+// Converts a PyObject * to a PyStackRef, stealing the reference.
+#if defined(Py_GIL_DISABLED)
+    static inline _PyStackRef
+    _PyStackRef_StealRef(PyObject *obj) {
+        // Make sure we don't take an already tagged value.
+        assert(PyStackRef_Get(((_PyStackRef){.bits = ((uintptr_t)(obj))})) == obj);
+        return ((_PyStackRef){.bits = ((uintptr_t)(obj))});
+    }
+    #define PyStackRef_StealRef(obj) _PyStackRef_StealRef(_PyObject_CAST(obj))
+#else
+    #define PyStackRef_StealRef(obj) ((_PyStackRef){.bits = ((uintptr_t)(obj))})
+#endif
+
+#if defined(Py_GIL_DISABLED)
+    static inline PyObject *
+    PyStackRef_StealObject(_PyStackRef tagged) {
+        if ((tagged.bits & Py_TAG_DEFERRED) == Py_TAG_DEFERRED) {
+            assert(_PyObject_HasDeferredRefcount(PyStackRef_Get(tagged)));
+            return Py_NewRef(PyStackRef_Get(tagged));
+        }
+        return PyStackRef_Get(tagged);
+    }
+#else
+    #define PyStackRef_StealObject(tagged) PyStackRef_Get(tagged)
+#endif
+
+static inline void
+_Py_untag_stack_borrowed(PyObject **dst, const _PyStackRef *src, size_t length) {
+    for (size_t i = 0; i < length; i++) {
+        dst[i] = PyStackRef_Get(src[i]);
+    }
+}
+
+static inline void
+_Py_untag_stack_steal(PyObject **dst, const _PyStackRef *src, size_t length) {
+    for (size_t i = 0; i < length; i++) {
+        dst[i] = PyStackRef_StealObject(src[i]);
+    }
+}
+
+
+#define PyStackRef_XSETREF(dst, src) \
+    do { \
+        _PyStackRef *_tmp_dst_ptr = _Py_CAST(_PyStackRef*, &(dst)); \
+        _PyStackRef _tmp_old_dst = (*_tmp_dst_ptr); \
+        *_tmp_dst_ptr = (src); \
+        PyStackRef_XDECREF(_tmp_old_dst); \
+    } while (0)
+
+#define PyStackRef_SETREF(dst, src) \
+    do { \
+        _PyStackRef *_tmp_dst_ptr = _Py_CAST(_PyStackRef*, &(dst)); \
+        _PyStackRef _tmp_old_dst = (*_tmp_dst_ptr); \
+        *_tmp_dst_ptr = (src); \
+        PyStackRef_DECREF(_tmp_old_dst); \
+    } while (0)
+
+#define PyStackRef_CLEAR(op) \
+    do { \
+        _PyStackRef *_tmp_op_ptr = _Py_CAST(_PyStackRef*, &(op)); \
+        _PyStackRef _tmp_old_op = (*_tmp_op_ptr); \
+        if (PyStackRef_Get(_tmp_old_op) != NULL) { \
+            *_tmp_op_ptr = PyStackRef_StealRef(_Py_NULL); \
+            PyStackRef_DECREF(_tmp_old_op); \
+        } \
+    } while (0)
+
+#if defined(Py_GIL_DISABLED)
+    static inline void
+    PyStackRef_DECREF(_PyStackRef tagged) {
+        if ((tagged.bits & Py_TAG_DEFERRED) == Py_TAG_DEFERRED) {
+            return;
+        }
+        Py_DECREF(PyStackRef_Get(tagged));
+    }
+#else
+    #define PyStackRef_DECREF(op) Py_DECREF(PyStackRef_Get(op))
+#endif
+
+#define PyStackRef_DECREF_OWNED(op) Py_DECREF(PyStackRef_Get(op));
```

Comment:
What is this?

Reply:
This made me think quite abit. I just realised my implementation of build slice is bugged in the failure case. Anything that should need something like this should be bugged. Thanks!

Reply:
Ok, but do we need it? If we do, what's the purpose of the macro? If not, let's delete it.

Reply:
No we dont if we manage our lifetimes properly. I will remove it.